rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java

```diff
@@ -1669,10 +1663,33 @@ public BooleanWrapper shutdown(boolean preempt) {
                 removeAllNodes(entry.getKey(), preempt, true);
                 entry.getValue().shutdown(caller);
             }
+            waitForAllNodeSourcesToBeShutdown();
+            finalizeShutdown();
         }
         return new BooleanWrapper(true);
     }
 
+    private void waitForAllNodeSourcesToBeShutdown() {
+        boolean atLeastOneAlive = false;
+        int millisBeforeHardShutdown = 0;
+        try {
+            do {
+                millisBeforeHardShutdown++;
+                Thread.sleep(100);
+                for (Entry<String, NodeSource> entry : this.deployedNodeSources.entrySet()) {
+                    try {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Extract this nested try block into a separate method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)